### PR TITLE
Issue 105: Add support for sa_key_provision_ta()

### PR DESCRIPTION
**Pull request by riwoh**

### CODE REVIEW

File: reference/src/client/include/sa_key.h [hunk 1]
```diff
@@ -93,6 +93,112 @@ sa_status sa_key_export(
...
 
+/* **TA Key Type Definition**
+
+The operator provisioning key type will be communicated through the following enumeration:
+*/
+typedef enum  {
+  WIDEVINE_OEM_PROVISIONING,
+  PLAYREADY_MODEL_PROVISIONING,
+  NETFLIX_PROVISIONING
+} sa_ta_key_type;
```
Comment (hchoi-brcm) [inline reference/src/client/include/sa_key.h:104]: Other types and enumeration definitions are in sa_types.h and sa_key.h contains function declarations for key module. Could we keep this concept for TA key provision?

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+               return status_hmac;
+            }
+            if(SA_STATUS_OK != status_wrapping) {
+               ERROR("Faild to import netflix wrapping key");
+            }
+            return status_wrapping;
+         }
+         break;
+      default:
+         break;
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:358]: We should report that the key provisioning type is unknown if we get here.

File: reference/src/client/include/sa_key.h [hunk 3]
```diff
@@ -93,6 +93,112 @@ sa_status sa_key_export(
...
         sa_key key);
 
+/* **TA Key Type Definition**
+
+The operator provisioning key type will be communicated through the following enumeration:
+*/
+typedef enum  {
+  WIDEVINE_OEM_PROVISIONING,
+  PLAYREADY_MODEL_PROVISIONING,
+  NETFLIX_PROVISIONING
```
Comment (mhabrat) [inline reference/src/client/include/sa_key.h:103]: Include support for Apple provisioning.

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+   PlayReadyProvisioning   *prProvision = NULL;
+   NetflixProvisioning     *nflxProvision = NULL;
+
+   if(NULL == in ||
+      0 == in_length) {
+      ERROR("null input or input size is  0");
+      return SA_STATUS_NULL_PARAMETER;
+   }
+
+   switch(ta_key_type) {
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:72]: Add support for Apple provisioning.

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+               0 == ((WidevineOemProvisioning*)in)->oemDevicePrivateKeyLength   ||
+               NULL == ((WidevineOemProvisioning*)in)->oemDeviceCertificate     ||
+               0 == ((WidevineOemProvisioning*)in)->oemDeviceCertificateLength) {
+               ERROR("null private key/certor or size is  0");
+               return SA_STATUS_NULL_PARAMETER;
+            }
+
+            wvProvision = calloc(sizeof(WidevineOemProvisioning), 1);
+            if(NULL == wvProvision) {
+               ERROR("OOM");
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:87]: Please adopt standard secapi3 error reporting.

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+               (void*)calloc(wvProvision->oemDevicePrivateKeyLength, 1);
+            if(NULL == wvProvision->oemDevicePrivateKey) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(wvProvision->oemDevicePrivateKey,
+                   ((WidevineOemProvisioning*)in)->oemDevicePrivateKey,
+                   wvProvision->oemDevicePrivateKeyLength);
+
+            /*here just provide an examlpe to pass certification data,
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:107]: Let's inlcude the "TODO SoC Vendor" label everywhere the SoC vendor is expected to make changes.

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+               wvProvision->oemDevicePrivateKeyLength -=8;
+            } else {
+              INFO("NO SIGN in the header");
+            }
+
+            sa_key *rsa_key = create_uninitialized_sa_key();
+            assert(NULL != rsa_key);
+            status = sa_key_import(rsa_key, SA_KEY_FORMAT_EXPORTED,
+                        wvProvision->oemDevicePrivateKey,
+                        wvProvision->oemDevicePrivateKeyLength,NULL);
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:135]: Shouldn't the parameters field provide the SA_KEY_FORMAT_SOC type to denote the key is protected with the SOC key?

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+            INFO("prProvision:%x",prProvision);
+            INFO("keyLen : %d", prProvision->privateKeyLength);
+            INFO("CertLen : %d", prProvision->modelCertificateLength);
+            INFO("Model type : %s",
+                  (prProvision->modelType == PLAYREADY_MODEL_2K)? "model 2000" :
+                  ((prProvision->modelType == PLAYREADY_MODEL_3K)? "model 3000":
+                   "No such model type"));
+
+            if(0 == (memcmp(prProvision->privateKey, "SIGN", 4))) {
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:214]: The SIGN attribute is specific to Widevine only.

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+               prProvision->privateKey += 8;
+               prProvision->privateKeyLength -=8;
+            } else {
+              INFO("NO SIGN in the header");
+            }
+
+            sa_key *rsa_key = create_uninitialized_sa_key();
+            assert(NULL != rsa_key);
+            status = sa_key_import(rsa_key, SA_KEY_FORMAT_EXPORTED,
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:223]: Shouldn't the parameters field provide the SA_KEY_FORMAT_SOC type to denote the key is protected with the SOC key?

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+               ERROR("Faild to import playready private key");
+            }
+            return status;
+         }
+         break;
+      case NETFLIX_PROVISIONING:
+         {
+            DEBUG("case NETFLIX_PROVISIONING");
+            INFO("hmac length: %d", ((NetflixProvisioning*)(in))->hmacKeyLength);
+            INFO("wraaping length: %d", ((NetflixProvisioning*)(in))->wrappingKeyLength);
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:250]:
```suggestion
            INFO("wraping length: %d", ((NetflixProvisioning*)(in))->wrappingKeyLength);
```

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+            INFO("nflxProvision:%x",nflxProvision);
+            INFO("hmacLen : %d", nflxProvision->hmacKeyLength);
+            INFO("wrappingLen : %d", nflxProvision->wrappingKeyLength);
+            INFO("ESNLen : %d", nflxProvision->esnContainerLength);
+
+            sa_key *sa_hmac_key = create_uninitialized_sa_key();
+            assert(NULL != sa_hmac_key);
+            sa_status status_hmac = sa_key_import(sa_hmac_key, SA_KEY_FORMAT_EXPORTED,
+                                  nflxProvision->hmacKey,nflxProvision->hmacKeyLength,
+                                  NULL);
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:319]: Shouldn't the parameters field provide the SA_KEY_FORMAT_SOC type to denote the key is protected with the SOC key?

File: reference/src/clientimpl/src/sa_key_provision.c [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+
+            sa_key *sa_wrapping_key = create_uninitialized_sa_key();
+            assert(NULL != sa_wrapping_key);
+
+            sa_status status_wrapping = sa_key_import(sa_wrapping_key, SA_KEY_FORMAT_EXPORTED,
```
Comment (mhabrat) [inline reference/src/clientimpl/src/sa_key_provision.c:324]: Shouldn't the parameters field provide the SA_KEY_FORMAT_SOC type to denote the key is protected with the SOC key?

File: reference/src/client/test/sa_key_provision.cpp [hunk 13]
```diff
@@ -0,0 +1,568 @@
...
+
+        return result;
+    }
+} // namespace
+
+#ifdef FILE_BASED_FETCH_KEY
+#include <sys/stat.h>
+#include <string.h>
+
+#define wievine_oem_privatekey "/opt/drm/0351000003510001.key"
```
Comment (mhabrat) [inline reference/src/client/test/sa_key_provision.cpp:439]: Let's not leak Comcast specific information via open source implementations.

File: reference/src/client/test/sa_key_provision.cpp [hunk 14]
```diff
@@ -0,0 +1,568 @@
...
+        auto header_wrapping_key = key_header(*sa_wrapping_key);
+        ASSERT_NE(nullptr, header_wrapping_key.get());
+        ASSERT_TRUE(memcmp(&rights, &header_wrapping_key->rights, sizeof(sa_rights)) == 0);
+        ASSERT_EQ(clear_wrapping_key.size(), header_wrapping_key->size);
+        ASSERT_EQ(SA_KEY_TYPE_SYMMETRIC, header_wrapping_key->type);
+    }
+#endif //KEY_VALIDATION_TEST
+
+#ifdef FILE_BASED_FETCH_KEY
+ /*there are two file 0351000003510001.key and 0651000006510001.bin under
```
Comment (mhabrat) [inline reference/src/client/test/sa_key_provision.cpp:386]: Let's not leak Comcast specific implementations through open source implementations.

**Review by mhabrat — CHANGES_REQUESTED** [review]

File: reference/src/client/include/sa_key.h [hunk 15]
```diff
@@ -93,6 +93,49 @@ sa_status sa_key_export(
...
+ * This function is proposed to support the provisioning of keys from an operator's key
+ * provisioning service to a specific TA.
+ * #1.	Field provisioning service delivers the encrypted TA Key to the device.
+ * #2.	The TA Key is imported and provisioned to SecAPI3 using new API: sa_key_provision_ta().
+ * #3.	SecAPI3 TA decrypts TA Key and converts key to SOC vendor TA Key format.
+ * #4.	SecAPI3 TA calls the SOC vendor-specific API to deliver the key to the TA.
+ * #5.	The key will be loaded for use within the TA. If the key received by the TA is new or updated
+        and the TA supports secure storage, it will store the key to TA Secure Storage.
+ * #6.	This provisioning flow (Steps 2-5) will occur upon every device reboot/initialization. SOC
+    Vendor TA Key updates will only be written to TAs supporting secure store upon receiving a new
```
Comment (mhabrat) [inline reference/src/client/include/sa_key.h:107]: Indentation.

File: reference/src/client/include/sa_key.h [hunk 16]
```diff
@@ -93,6 +93,49 @@ sa_status sa_key_export(
...
+    or updated key. The TA Key is not imported/stored in SecAPI3 when the sa_key_provision_ta() API
+    is used.
+ *
+ * @param[in] ta_key_type (enum) specifies the type of key being provisioned to the TA as defined in
+    [TA Key Type Definition](#ta-key-type-definition).
```
Comment (mhabrat) [inline reference/src/client/include/sa_key.h:112]:
```suggestion
   [TA Key Type Definition](#ta-key-type-definition).
```

File: reference/src/client/include/sa_types.h [hunk 17]
```diff
@@ -1030,6 +1032,93 @@ typedef struct {
...
+  APPLE_FAIRPLAY_PROVISIONING,
+  NETFLIX_PROVISIONING
+} sa_key_type_ta;
+
+/** PlayReady model types */
+
+typedef enum {
+  PLAYREADY_MODEL_2K,
+  PLAYREADY_MODEL_3K
+}PLAYREADY_MODEL_TYPE;
```
Comment (mhabrat) [inline reference/src/client/include/sa_types.h:1052]:
```suggestion
} PLAYREADY_MODEL_TYPE;
```

File: reference/src/client/test/sa_crypto_cipher_process_rsa_pkcs1v15.cpp [hunk 18]
```diff
@@ -133,7 +133,7 @@ namespace {
 
         ASSERT_EQ(status, SA_STATUS_OK);
         ASSERT_NE(cipher, nullptr);
-
+
```
Comment (mhabrat) [inline reference/src/client/test/sa_crypto_cipher_process_rsa_pkcs1v15.cpp:136]: Remove excess blank space.

File: reference/test/SecApiKeyProvisionTaTests.md [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+
+## Placing Files in /keys Directory:
+
+Move Key Provision Files: Copy the key provision files you want to use for testing to the /keys
+directory in your project or system.
+
+Enable Flag in sa_key_import_common.h: Similar to method 1, edit the sa_key_import_common.h header
+file and turn on the FILE_BASED_FETCH_KEY flag. This enables the test fucntion to locate the key
+provision files from the /keys directory.
+
```
Comment (mhabrat) [inline reference/test/SecApiKeyProvisionTaTests.md:44]: Remove excess blank lines at endo of file.

**Review by mhabrat — CHANGES_REQUESTED** [review]

**Review by mhabrat — APPROVED** [review]

**Pull request merged**